helix-term/src/commands/typed.rs

```diff
@@ -913,6 +913,29 @@ fn setting(
     Ok(())
 }
 
+/// Change the language of the current buffer at runtime.
+fn language(
+    cx: &mut compositor::Context,
+    args: &[Cow<str>],
+    _event: PromptEvent,
+) -> anyhow::Result<()> {
+    if args.len() != 1 {
+        anyhow::bail!("Bad arguments. Usage: `:set-language language`");
+    }
+
+    let doc = doc_mut!(cx.editor);
+    let loader = Some(cx.editor.syn_loader.clone());
+    let config = cx.editor.syn_loader.language_configs().find_map(|config| {
+        if config.language_id == args[0] {
+            Some(config.clone())
+        } else {
+            None
+        }
+    });
+    doc.set_language(config, loader);
```

Comment:
Similar to https://github.com/helix-editor/helix/pull/1787 you also need to call refresh_language_server to change language servers. The function also needs to be fixed: it the server changes we need to send a text document close to the old one.

Reply:
The text document close seems to be already handled in `Editor::launch_language_server()`, more specifically from L464:
```rust
        if let Some(language_server) = language_server {
            // only spawn a new lang server if the servers aren't the same
            if Some(language_server.id()) != doc.language_server().map(|server| server.id()) {
                if let Some(language_server) = doc.language_server() {
                    tokio::spawn(language_server.text_document_did_close(doc.identifier()));
                }

                let language_id = doc.language_id().map(ToOwned::to_owned).unwrap_or_default();

                // TODO: this now races with on_init code if the init happens too quickly
                tokio::spawn(language_server.text_document_did_open(
                    doc.url().unwrap(),
                    doc.version(),
                    doc.text(),
                    language_id,
                ));

                doc.set_language_server(Some(language_server));
            }
        }
```
The only fixing that's needed for `Editor::refresh_language_server()` is to check if user already set a language or not, otherwise the detection will overwrite the user set value.
```diff
     pub fn refresh_language_server(&mut self, doc_id: DocumentId) -> Option<()> {
         let doc = self.documents.get_mut(&doc_id)?;
-        doc.detect_language(self.syn_loader.clone());
+        // try detection only if language is not already set for this document.
+        if doc.language.is_none() {
+            doc.detect_language(self.syn_loader.clone());
+        }
         Self::launch_language_server(&mut self.language_servers, doc)
     }
```

I'm not sure if this check should be part of `Document::detect_language()` or not.